test/restrictions.js

```diff
@@ -0,0 +1,38 @@
+require("should");
+const path = require("path");
+const fs = require("fs");
+const ResolverFactory = require("../lib/ResolverFactory");
+const CachedInputFileSystem = require("../lib/CachedInputFileSystem");
+
+const fixture = path.resolve(__dirname, "fixtures", "restrictions");
+const nodeFileSystem = new CachedInputFileSystem(fs, 4000);
+
+it("should respect RegExp restriction", done => {
+	const resolver = ResolverFactory.createResolver({
+		extensions: [".js"],
+		fileSystem: nodeFileSystem,
+		restrictions: [/\.(sass|scss|css)$/]
+	});
+
+	resolver.resolve({}, fixture, "pck1", {}, (err, result) => {
+		if (!err) throw new Error(`expect error, got ${result}`);
+		err.should.be.instanceof(Error);
+		err.message.should.match(/Resolve restriction \/.*\/ not passed/);
+		done();
+	});
+});
+
+it("should respect string restriction", done => {
+	const resolver = ResolverFactory.createResolver({
+		extensions: [".js"],
+		fileSystem: nodeFileSystem,
+		restrictions: [fixture]
+	});
+
+	resolver.resolve({}, fixture, "pck2", {}, (err, result) => {
+		if (!err) throw new Error(`expect error, got ${result}`);
+		err.should.be.instanceof(Error);
+		err.message.should.match(/Resolve restriction ".*" not passed/);
+		done();
+	});
+});
```

Comment:
Can we add tests (with the `restrictions` option) where:
- we have `.css` in `extensions` - should resolve `index.css`
- we have `style` in `mainFields` - should resolve `index.css`
- we have `index.css` - `package/index.css`, `package/index.js`, `package.json` with `main: 'index.js'` (`mainFiles`) - should resolve `index.css`

Reply:
I think to get the expected behavior (resolving should try alternatives when hitting a restriction) the restrictions plugin need to return with `null` instead of throwing an error. An error would cancel the full resolving.

You can "log" (resolveContext.log) an message that a restriction was hit.

Reply:
Ah, I misunderstand expected behavior, fixed